pytensor/tensor/subtensor.py

```diff
@@ -197,18 +269,47 @@ def get_canonical_form_slice(
     in which 0 <= start <= stop <= length and step > 0, and a flag which says
     if the resulting set of numbers needs to be reversed or not.
 
+    Given a scalar index `idx` that may or not be negative, convert it to
+    a certainly positive form `idx if idx >= 0 else length + idx`.
+
+    Returns
+    -------
+    slc
+        Canonical form slice or scalar variable.
+    direction
+        Direction to iterate the resulting elements in. (-1 or 1). May be symbolic.
     """
     from pytensor.tensor import ge, lt, sign, switch
 
-    if not isinstance(theslice, slice):
-        try:
-            value = as_index_literal(theslice)
-        except NotScalarConstantError:
-            value = theslice
-
-        value = switch(lt(value, 0), (value + length), value)
+    # Convert the two symbolic slice types into a native slice
+    if isinstance(theslice, SliceConstant):
+        theslice = cast(slice, theslice.data)
+    elif isinstance(theslice, Variable) and isinstance(theslice.type, SliceType):
+        theslice = slice(*theslice.owner.inputs)
+    # Other non-slice types are the scalar indexing case
+    elif not isinstance(theslice, slice):
+        sslice: int | np.integer | ScalarVariable | TensorVariable
+        if isinstance(theslice, ScalarVariable):
+            sslice = theslice
+        elif isinstance(theslice, TensorVariable) and theslice.ndim == 0:
+            sslice = theslice
+        else:
```

Comment:
Just raise a TypeError, no other types should show up

Reply:
There are almost a hundred places where `NotScalarConstantError` is caught. Replacing it with a generic `TypeError` appears very risky..

Reply:
It's being misused, why misuse it even more?